File: 20221028_Friday/20221028.rb
# Leetcode: 49. Group Anagrams.
# https://leetcode.com/problems/group-anagrams/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 533 ms, faster than 20.43% of Ruby online submissions for Group Anagrams.
# Memory Usage: 215.4 MB, less than 61.65% of Ruby online submissions for Group Anagrams.
# 2022.10.28 Daily Challenge.
# @param {String[]} strs
# @return {String[][]}
def group_anagrams(strs)
  hash = {}
  strs.each do |str|
    key = get_hash_map(str)
    hash[key] ||= []
    hash[key].push(str)
  end
  hash.values
end

def get_hash_map(str)
  ans = Array.new(26,0)
  str.chars.each do |x|
    ans[x.ord - ?a.ord] += 1
  end
  return ans
end

File: 20221028_Friday/20221028.go
// Leetcode: 49. Group Anagrams.
// https://leetcode.com/problems/group-anagrams/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 74 ms, faster than 16.18% of Go online submissions for Group Anagrams.
// Memory Usage: 9.5 MB, less than 9.79% of Go online submissions for Group Anagrams.
// 2022.10.28 Daily Challenge.
import "strings"
import "strconv"
func groupAnagrams(strs []string) [][]string {
  hash := make(map[string][]string)
  for _, str := range strs{
    key := get_key(str)
    arr, exists := hash[key]
    if ! exists {
      hash[key] = make([]string,0)
    }
    hash[key] = append(arr,str)
  }
  answer := make ([][]string,0)
  for _, val := range hash {
    answer = append(answer, val)
  }
  return answer
}

func get_key(str string) string{
  counter := make([]int,26)
  for _,char := range str {
    counter[int(char) - 97] += 1
  }
  answer := make([]string,len(counter))
  for i, number := range counter{
    answer[i] = strconv.Itoa(number)
  }
  return strings.Join(answer,".")
}
